### PR TITLE
DM-18861: Enhance "how to write a SAL script" documentation

### PR DESCRIPTION
**Pull request by r-owen**

### CODE REVIEW

File: doc/sal_scripts.rst [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+---------------------------
+
+If your script can be configured, define the configure method::
+
+    def configure(self, param1, param2=default, ...):
+        # ...
+
+This method should verify that all parameters have usable type and reasonable values.
+This method can then set attributes or perform actions accordingly.
+However, if at all possible please make it synchronous and reasonably quick.
```
Comment (pi11235) [inline doc/sal_scripts.rst]: For my education, why must it be synchronous?

File: doc/sal_scripts.rst [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+Note that ``configure`` cannot be called once the script is running.
+However, `BaseScript` enforces this automatically; you don't need to do anything to prevent it.
+
+cleanup method (optional)
+-------------------------
+
+When your script is ending, after ``run`` finishes, is stopped early, or raises an exception, ``BaseScript`` calls ``cleanup`` for final cleanup.
```
Comment (pi11235) [inline doc/sal_scripts.rst]: Great, but what does cleanup do? When might someone use this?

File: doc/sal_scripts.rst [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+In some sense ``configure`` is like the ``finally`` clause of a ``try/finally`` block.
```
Comment (pi11235) [inline doc/sal_scripts.rst]: you mean cleanup, not configure?

File: python/lsst/ts/scriptqueue/base_script.py [hunk 4]
```diff
@@ -336,13 +336,13 @@ def assert_state(self, action, states):
         * Receive the configuration as a ``yaml`` string.
```
Comment (pi11235) [inline python/lsst/ts/scriptqueue/base_script.py]: Do we have documentation about how we do YAML files yet? If so, we should put links in the appropriate places

File: doc/sal_scripts.rst [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+.. _lsst.ts.scriptqueue_sal_scripts:
+
+SAL Scripts
+###########
+
+SAL scripts implement one-off tasks: they are configured once, run once and then they quit.
+Thus they are very different from Commandable SAL Components (CSCs), which typically run indefinitely.
+
+A SAL script is a SAL component that supports the ``Script`` API and follows these rules:
```
[the remark this answers is not in the document]
Comment (r-owen) [reply doc/sal_scripts.rst]: What I'm trying to say is that any executable SAL component that follows the listed rules can be run as a SAL script by the Script Queue. I hope we will only have SAL scripts that are written in Python, but I did not want to needlessly prohibit scripts in other languages. The

File: doc/sal_scripts.rst [hunk 6]
```diff
@@ -0,0 +1,135 @@
+.. py:currentmodule:: lsst.ts.scriptqueue
+
+.. _lsst.ts.scriptqueue_sal_scripts:
+
+SAL Scripts
+###########
+
+SAL scripts implement one-off tasks: they are configured once, run once and then they quit.
+Thus they are very different from Commandable SAL Components (CSCs), which typically run indefinitely.
```
[the remark this answers is not in the document]
Comment (r-owen) [reply doc/sal_scripts.rst]: Coordination is a good concept. I'll try to work that in.

I agree that a CSC does not complete -- that's what I meant by "run indefinitely".

File: doc/index.rst [hunk 7]
```diff
@@ -13,124 +13,12 @@ Suppport for running SAL scripts on a script queue.
 Using lsst.ts.scriptqueue
```
[the remark this answers is not in the document]
Comment (r-owen) [reply doc/index.rst]: Good idea. I have added that.

File: doc/index.rst [hunk 8]
```diff
@@ -13,124 +13,12 @@ Suppport for running SAL scripts on a script queue.
...
 =========================
```
[the remark this answers is not in the document]
Comment (r-owen) [reply doc/index.rst]: Yes. The point of "SAL" in "SAL Scripts" is to differentiate from any other kind of script. SAL Scripts are managed by the ScriptQueue (which maybe should have been called the SALScriptQueue but I don't propose to change that name) and do all their communicating via SAL.

File: doc/sal_scripts.rst [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+
+Script Packages
+###############
+
+All SAL scripts should go into the ``scripts`` directory of one of the following packages, so the `ScriptQueue` can find them:
+
+* ``ts_standardscripts``: scripts that are approved for regular use.
+* ``ts_externalscripts``: scripts for experimentation and one-off tasks.
+
```
[the remark this answers is not in the document]
Comment (r-owen) [reply doc/sal_scripts.rst]: Yes. Users *can* create their own dir. I also fear chaos but I think we need procedures to manage it and until we have them I can't document them.

File: doc/sal_scripts.rst [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+.. _lsst.ts.scriptqueue_python_sal_scripts:
+
+Python SAL Scripts
+##################
+
+Each SAL Script written in Python should consist of two parts:
+
+* The script file itself, as an executable file in the ``scripts/...`` subdirectory of the ``ts_standardscripts`` or ``ts_externalscripts`` package.
+* An implementation in the ``python/lsst/ts/...`` subdirectory of the same package.
```
[the remark this answers is not in the document]
Comment (r-owen) [reply doc/sal_scripts.rst]: I have expanded this. But if we have non-programmers writing these scripts I am worried.

File: doc/sal_scripts.rst [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+  Technically the implementation can be in the script file, but that makes the code harder to reuse and harder to find.
+
+The script file is trivial and will look something like this::
```
[the remark this answers is not in the document]
Comment (r-owen) [reply doc/sal_scripts.rst]: OK, but it really should only be 3 lines of code -- exactly like the 3 shown, with the names changed.

File: doc/sal_scripts.rst [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+run method (required)
+---------------------
+
+Define asynchronous function to implement the main functionality of your script::
+
+    async def run(self):
+        # ...
+
+If ``run`` needs to run a slow computation, either call ``await asyncio.sleep(0)`` occasionally to give other coroutines a chance to run, or else run the computation in a thread using `run_in_executor`_ e.g.::
```
[the remark this answers is not in the document]
Comment (r-owen) [reply doc/sal_scripts.rst]: Yes I do. That frees the event loop. I"ll add a note.

File: doc/sal_scripts.rst [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    # do other work here...
+    # then eventually you must wait for the background task
+    result = await thread_task
+
+.. _run_in_executor: https://docs.python.org/3/library/asyncio-eventloop.html#id14
+
+checkpoints
+^^^^^^^^^^^
+
+In your run method you may call ``await self.checkpoint([name])`` to specify a point at which users can pause or stop the script.
```
[the remark this answers is not in the document]
Comment (r-owen) [reply doc/sal_scripts.rst]: I'll use a variable named `name_of_checkpoint` and remove the `[]`. I really want to encourage people to provide a name.

File: doc/sal_scripts.rst [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+By providing a name for each checkpoint you allow users to specify exactly where they would like the script to pause or stop.
+In addition, each checkpoint is reported as the ``last_checkpoint`` attribute of the ``state`` event, so providing checkpoints with informative names can be helpful in tracking the progress of a script.
+We suggest you make checkpoint names fairly short, obvious and unique, but none of these rules is enforced.
+If you have a checkpoint in a loop you may wish to modify the name for each iteration, e.g.::
+
+    for iter in range(num_iter):
+        await self.checkpoint(f"start_loop{iter}")
+
+This allows the user to pause or stop at any particular iteration, and makes the reported state more informative.
```
[the remark this answers is not in the document]
Comment (r-owen) [reply doc/sal_scripts.rst]: I'll change that to

    makes the ``state`` event more informative

File: python/lsst/ts/scriptqueue/base_script.py [hunk 15]
```diff
@@ -76,7 +76,7 @@ class BaseScript(salobj.Controller, abc.ABC):
     remotes_dict : `dict` of `str` : `salobj.Remote` (optional)
         Dict of attribute name: `salobj.Remote`, or `None` if no remotes.
         These remotes are added as attributes of ``self`` and are also
-        used to generate a list of remote names for script metadata.
+        used to generate a list of remote names for ``description`` event.
```
[the remark this answers is not in the document]
Comment (r-owen) [reply python/lsst/ts/scriptqueue/base_script.py]: Fair enough. I'll just make it a reference to the main description.

File: python/lsst/ts/scriptqueue/base_script.py (hunk 4 above)
Comment (r-owen) [reply python/lsst/ts/scriptqueue/base_script.py]: Good question. These won't be necessarily be files (unlike CSC config), and they need to decode as dicts. I'll say something.

**Pull request merged**